### ISSUE TITLE
overly broad typevar matching against a union parameter with multiple typevar occurrences

### ISSUE THREAD

**Issue by marovira** [astral-sh/ty#2870]
### Question

I'm exploring the option of migrating one of my projects to `ty` and came across this issue that I'm not sure if it's an actual bug or I'm missing something on how `ty` expects me to deal with it.
Consider this simplified example:
```python
# Save to a file called list.py
import typing

T = typing.TypeVar("T")

def convert_to_list(var: T | list[T] | tuple[T, ...]) -> list[T]:
    if isinstance(var, list):
        return var
    if isinstance(var, tuple):
        return list(var)
    return [var]

# Type hints here are to keep mypy happy and show what I'm expecting the return to be.
a = convert_to_list(1) # A
b: list[int] = convert_to_list([1, 2, 3]) # B
c: list[int] = convert_to_list((1, 2, 3)) # C
```
If I run `mypy` or `zuban` on this, I get no errors. If I use `ty check list.py`, then I get the following output:
```
error[invalid-argument-type]: Argument to bound method `__init__` is incorrect
  --> list.py:9:21
   |
 7 |         return var
 8 |     if isinstance(var, tuple):
 9 |         return list(var)
   |                     ^^^ Expected `Iterable[T@convert_to_list]`, found `(T@convert_to_list & tuple[object, ...]) | tuple[T@convert_to_list, ...]`
10 |     return [var]
   |
info: Element `T@convert_to_list & tuple[object, ...]` of this union is not assignable to `Iterable[T@convert_to_list]`
info: Matching overload defined here
    --> stdlib\builtins.pyi:2839:9
     |
2837 |     def __init__(self) -> None: ...
2838 |     @overload
2839 |     def __init__(self, iterable: Iterable[_T], /) -> None: ...
     |         ^^^^^^^^       ---------------------- Parameter declared here
2840 |     def copy(self) -> list[_T]:
2841 |         """Return a shallow copy of the list."""
     |
info: Non-matching overloads for bound method `__init__`:
info:   (self) -> None
info: rule `invalid-argument-type` is enabled by default

error[invalid-assignment]: Object of type `list[list[int] | int]` is not assignable to `list[int]`
  --> list.py:13:4
   |
12 | a = convert_to_list(1)
13 | b: list[int] = convert_to_list([1, 2, 3])
   |    ---------   ^^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `list[list[int] | int]`
   |    |
   |    Declared type
14 | c: list[int] = convert_to_list((1, 2, 3))
   |
info: rule `invalid-assignment` is enabled by default

error[invalid-assignment]: Object of type `list[tuple[int, int, int] | int]` is not assignable to `list[int]`
  --> list.py:14:4
   |
12 | a = convert_to_list(1)
13 | b: list[int] = convert_to_list([1, 2, 3])
14 | c: list[int] = convert_to_list((1, 2, 3))
   |    ---------   ^^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `list[tuple[int, int, int] | int]`
   |    |
   |    Declared type
   |
info: rule `invalid-assignment` is enabled by default

Found 3 diagnostics
```

There's two parts that puzzle me about this:
1. In the `if isinstance(var, tuple)`, Ty is inferring the type to be either `T & tuple[object]` or `tuple[T, ...]`. According to the info tag of [this](https://docs.astral.sh/ty/features/type-system/#top-and-bottom-materializations) section, I could perform the `isinstance` check against `T` to rule out the intersection, but obviously that won't work here since we're dealing with a generic. The question then becomes: how am I supposed to deal with this error? Based on the code, I'm really struggling to understand why Ty believes that the argument into the `list` constructor is invalid given that I've already checked that the type of `var` is `tuple`, which is a valid type into the `list` constructor. 
2. The return type for cases `B` and `C` is really confusing me. In case B, I have no idea how Ty arrives at the conclusion that the function can return a list of lists or an int since it would clearly fall under the `isinstance(var, list)` check and would therefore just return a `list[int]`. As for case C, I guess I can kind of see how Ty decided that the return is a list of tuples, but it doesn't make sense based on the actual code and the checks I'm making.

Is this an actual bug I stumbled upon, or does Ty's logic exceeds my understanding?

### Version

ty 0.0.18 (751672710 2026-02-20)

**Comment by carljm** [comment]
Hi, thanks for the very clear report!

Issue (1) is the same as #2843, so let's track it over there. (To answer your question about the error on `list(var)`: we are taking into account that this is a return, and the expected return type is `list[T]`, so we are handling this call with that constraint in mind, and erroring because we can't prove the list will be a `list[T]`. The `isinstance(var, tuple)` check does not inherently rule out the possibility of `T` being a tuple of something other than `T`, and that's why we error. Except that in this case, we should consider that any tuple argument would have matched against the `tuple[T, ...]` arm in the parameter type, so we _should_ be able to conclude that any tuple must be a tuple of `T`.)

I think issue (2) is a limitation in ty's generic solver that we aim to fix. Let's keep this issue open to track that specifically (the call-site inference). I think it's a sub-issue of #2799.

When resolving a call to a generic function, no type checker takes into account the body of the function -- it is resolved solely from the signature annotations. What ty is doing wrong here is that when we match e.g. the argument `[1, 2, 3]` against `T | list[T] | tuple[T, ...]` parameter type, we should say "this is a list, therefore let me prefer `list[T]`". Instead we match the list against both `T` and `list[T]`, and resolve the return type accounting for both possibilities (which is how you get the nested list in the return type). So this part is not related to the code in the body of the function, just that we are being too broad in how we match the typevar in the parameter annotation to the argument.

**Comment by marovira** [comment]
Ohhh so that's what's going on. It's definitely interesting how type deduction work for Python. Most of my experience comes from C++ and some Rust, which is why I loosely equated the arms of the if statements into an `if constexpr` and was confused by the errors `ty` was giving. Thanks for the explanation!

**Comment by carljm** [comment]
Here's a similar example, where we currently error on the last line because we think the `st.map` call returns `Streamable[Awaitable[str] | str]`, where we should infer just `Streamable[str]`, because we should prefer to match the signature of `call` against just `AsyncMapper`, not also against `Mapper`:

```py
from typing import AsyncIterable, Awaitable, Callable, Protocol

type SyncMapper[T, U] = Callable[[T], U]
type AsyncMapper[T, U] = Callable[[T], Awaitable[U]]
type Mapper[T, U] = SyncMapper[T, U] | AsyncMapper[T, U]

class Streamable[T](AsyncIterable[T]):
    def map[U](self, func: Mapper[T, U]) -> Streamable[U]:
        raise NotImplementedError

async def call(input: str) -> str:
    return ""

async def check() -> AsyncIterable[str]:
    st : Streamable[str] = Streamable()
    return st.map(call)
```

**Comment by carljm** [comment]
https://github.com/astral-sh/ty/issues/3746#issuecomment-4685682662 has a case where this issue occurs in matching a callable against `Callable[P, R] | Callable[Concatenate[type[T], P], R]`, where the preferable behavior is to match the second arm of the union, if the first parameter of the callable matches `type[T]`, but we currently just match `Callable[P, R]`.
